### PR TITLE
Added MessageHub polling to refresh Elasticsearch cache

### PR DESCRIPTION
**Pull request by fabiogomezdiaz**
- Added MessageHub topic subscription and message polling capability  …
- If reload_cache message found, then loads updated database into Elasticsearch
- Added configuration file to put MessageHub and Elasticsearch credentials and run locally
- Added scripts to load MessageHub and Elasticsearch credentials into environment variables
- Added checks for Message Hub and Elasticsearch credentials before app starts polling
- Updated gradle docker task and startup.sh script to load configuration into docker
- Updated dockerfile to copy config.json and load_config.sh files
- Updated docks to run application locally, in docker, and in Bluemix
- TODO: Flow diagram and Elasticsearch container deployment instructions

### CODE REVIEW

File: config.json [hunk 1]
```diff
@@ -0,0 +1,22 @@
+{
+  "es_connection_string": "PASTE_ELASTIC_SEARCH_CONNECTION_STRING_HERE",
+  "es_index": "api",
+  "es_doc_type": "items",
+  "mh_topic": "inventory",
+  "mh_message": "refresh_cache",
+  "VCAP_SERVICES": {
+    "messagehub": [{
+      "credentials": PASTE_MESSAGE_HUB_CREDENTIALS_OBJECT_HERE,
```
Comment (fabiogomezdiaz) [inline config.json]: This is intentional so you just copy and paste from Message Hub Bluemix credentials

File: docker/Dockerfile [hunk 2]
```diff
@@ -14,6 +14,10 @@ RUN apt-get update && apt-get install -y jq
 # Copy agent files
 COPY agents/ /agents/
 
+# Copy configuration files
```
Comment (fabiogomezdiaz) [inline docker/Dockerfile]: Added this to make it easy to run it in docker

File: docker/startup.sh [hunk 3]
```diff
@@ -6,6 +6,7 @@ export JAVA_OPTS="-Djava.security.egd=file:/dev/./urandom"
 
 # Load agent support if required
 source ./agents/newrelic.sh
+source load_config.sh
```
Comment (fabiogomezdiaz) [inline docker/startup.sh]: Complements the above

File: load_config.sh [hunk 4]
```diff
@@ -0,0 +1,18 @@
+#!/usr/bin/env bash
```
Comment (fabiogomezdiaz) [inline load_config.sh]: To make it easy to load MessageHub and Elasticsearch credentials and run locally and in docker

File: src/main/java/inventory/mysql/Application.java [hunk 5]
```diff
@@ -15,5 +15,26 @@
     public static void main(String[] args) {
         ApplicationContext ctx = SpringApplication.run(Application.class, args);
         System.out.println("Inventory microservice is ready for business...");
+
+        String connection = System.getenv("es_connection_string");
```
Comment (fabiogomezdiaz) [inline src/main/java/inventory/mysql/Application.java]: Checks to make sure we have what we need before starting

File: src/main/java/inventory/mysql/ElasticSearch.java [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+    }
+
+    // Get all rows from database
+    public JSONArray get_all_rows() {
+        JSONArray rows = null;
+        try {
+            // TODO: Use some SpringBoot data approach to get the data
+            OkHttpClient client = new OkHttpClient();
+            Request request = new Request.Builder()
+            .url("http://localhost:8080/micro/inventory")
```
Comment (fabiogomezdiaz) [inline src/main/java/inventory/mysql/ElasticSearch.java]: I know! Best I could come up with in the meantime

File: src/main/java/inventory/mysql/ElasticSearch.java [hunk 7]
```diff
@@ -0,0 +1,94 @@
+package inventory.mysql;
```
Comment (fabiogomezdiaz) [inline src/main/java/inventory/mysql/ElasticSearch.java:1]: This file will get all the rows from database and update the elasticsearch cache

File: src/main/java/inventory/mysql/MHConsumer.java [hunk 8]
```diff
@@ -0,0 +1,176 @@
+package inventory.mysql;
```
Comment (fabiogomezdiaz) [inline src/main/java/inventory/mysql/MHConsumer.java:1]: This file setups the JAAS configuration before talking with MessageHub. Then it starts polling topic

File: src/main/resources/application.yml [hunk 9]
```diff
@@ -3,7 +3,7 @@ eureka:
      preferIpAddress: true
```
Comment (fabiogomezdiaz) [inline src/main/resources/application.yml]: Oops, need to take this out

**Comment by gangchen03** [comment]
01/31 Review:
1. API definition of Inventory needs refinement, for example shouldn't have /update, /delete in REST path.
2. pre-requisite, for Docker, we shouldn't use the Docker-machine any more, use the Docker 1.12 or later. Use this link instead https://docs.docker.com/engine/installation/
3. ElasticSearch step 3, Copy the following as your ElasticSearch connection string what does it mean? Please explain, for example, the endpoint to communicate with ElasticSearch REST API?
4. MySQL load-data.sh script failed the first time, maybe add instruction to wait for MySQL fully started, 
5. Why is there a mysql folder under the micro-inventory project?
6. Run Inventory Service application on Localhost, please indicate which directory reader will be working under and,  give a complete example of the config.json
7. Validate the Running app locally, clearly mentioned which is the command, which is the result. They are now seems to be a single command.
8. Getting MessageHub Error
9. Config.json should be merged with application.yml file

**Comment by jkwong888** [comment]
My comments:
- I prefer to put the pre-requisites up-front, i.e. provision elasticsearch, mysql, and messagehub before starting on the build/deployment steps
- I prefer to not use that mysql repo.  The repo doesn't really add much value, as it's just the mysql image from docker hub with an extra script for the schema/data load.  Can we just document this instead:
  -  a docker pull mysql:latest from dockerhub
  - a docker tag and docker push into bluemix container image registry
  - cf ic run the mysql container
  - cf ic cp the script into the container
  - cf ic exec the script in the container

we can use compose for mysql (when it comes out of beta), or an on-prem mysql instance in production

Can we use the schema from the mysql repo and then deprecate or retire the mysql repo?

**Pull request merged**